adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java

```diff
@@ -183,66 +184,92 @@ private void handleEndpointConnection(final MqttEndpoint endpoint) {
+                LOG.debug("connection closed with client [{}]", endpoint.clientIdentifier());
+                if (registrationAssertions.remove(endpoint) != null)
+                    LOG.trace("removed registration assertion for client [{}]", endpoint.clientIdentifier());
+            });
 
-                try {
+            if (getConfig().isAuthenticationRequired()) {
 
-                    final ResourceIdentifier resource = ResourceIdentifier.fromString(message.topicName());
+                // check credentials for valid authentication
+                // so far, only hashed-password supported, more to follow
+                try {
 
-                    // if MQTT client doesn't specify device_id then closing connection (MQTT has now way for errors)
-                    if (resource.getResourceId() == null) {
-                        close(endpoint);
-                    } else {
+                    MqttUsernamePassword authObject = MqttUsernamePassword.create(endpoint,
+                            getConfig().isSingleTenant());
+
+                    validateCredentialsForDevice(authObject.getTenantId(), authObject.getType(), authObject.getAuthId(),
+                            authObject.getPassword()).setHandler(attempt -> {
+                                if (attempt.succeeded()) {
+                                    String deviceId = attempt.result();
+                                    LOG.trace("successfully authenticated device id <{}>", deviceId);
+                                    endpoint.accept(false);
+                                } else {
+                                    endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+                                }
+                    });
+                } catch (IllegalArgumentException e) {
+                    LOG.warn(e.getMessage());
+                    endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_BAD_USER_NAME_OR_PASSWORD);
+                }
+            } else {
+                endpoint.accept(false);
+            }
+        }
+    }
 
-                        Future<Void> messageTracker = Future.future();
-                        messageTracker.setHandler(s -> {
-                            if (s.failed()) {
-                                LOG.debug("cannot process message [client ID: {}, topic: {}, QoS: {}]: {}", endpoint.clientIdentifier(),
-                                        resource, message.qosLevel(), s.cause().getMessage());
-                                close(endpoint);
-                            } else {
-                                LOG.trace("successfully processed message [client ID: {}, topic: {}, QoS: {}]", endpoint.clientIdentifier(),
-                                        resource, message.qosLevel());
-                            }
-                        });
+    private Handler<MqttPublishMessage> getMqttPublishMessageHandlerImpl(final MqttEndpoint endpoint) {
```

Comment:
my point was to turn this method into the message handler itself, i.e. changing its signature to
`private void handleMessage(final MqttEndpoint endpoint, final MqttPublishMessage message)`
and then register it as a handler in the *handleEndpointConnection* method as follows:
````
...
  } else {
    endpoint.publishHandler(message -> handleMessage(endpoint, message));
    ....
````